core/src/main/java/org/apache/iceberg/BaseTableScan.java

```diff
@@ -167,59 +154,52 @@ public TableScan asOfTime(long timestampMillis) {
   @Override
   public TableScan option(String property, String value) {
     ImmutableMap.Builder<String, String> builder = ImmutableMap.builder();
-    builder.putAll(options);
+    builder.putAll(context.options());
     builder.put(property, value);
 
     return newRefinedScan(
-        ops, table, snapshotId, schema, rowFilter, ignoreResiduals,
-        caseSensitive, colStats, selectedColumns, builder.build());
+        ops, table, schema, context.options(builder.build()));
   }
 
   @Override
   public TableScan project(Schema projectedSchema) {
     return newRefinedScan(
-        ops, table, snapshotId, projectedSchema, rowFilter, ignoreResiduals,
-        caseSensitive, colStats, selectedColumns, options);
+        ops, table, projectedSchema, context.copy());
   }
 
   @Override
   public TableScan caseSensitive(boolean scanCaseSensitive) {
     return newRefinedScan(
-        ops, table, snapshotId, schema, rowFilter, ignoreResiduals,
-        scanCaseSensitive, colStats, selectedColumns, options);
+        ops, table, schema, context.caseSensitive(scanCaseSensitive));
   }
 
   @Override
   public TableScan includeColumnStats() {
     return newRefinedScan(
-        ops, table, snapshotId, schema, rowFilter, ignoreResiduals,
-        caseSensitive, true, selectedColumns, options);
+        ops, table, schema, context.colStats(true));
   }
 
   @Override
   public TableScan select(Collection<String> columns) {
     return newRefinedScan(
-        ops, table, snapshotId, schema, rowFilter, ignoreResiduals,
-        caseSensitive, colStats, columns, options);
+        ops, table, schema, context.selectedColumns(columns));
```

Comment:
I think it is worth renaming some of these methods; `selectedColumns` is a noun, not a verb. I would expect a method to be phrased using a verb, like `selectColumns(columns)`. Otherwise, this is just dropping `set` from a setter method.